### PR TITLE
VS2-Abstract sensor & path class

### PR DESCRIPTION
**Pull request by Michelbaartman**
https://trello.com/c/fcLwPy7y/13-vs2-abstract-sensor-path-class

### CODE REVIEW

File: BuildModule.cmake [hunk 1]
```diff
@@ -23,7 +23,7 @@ set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${cxxflags}")
 set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -nostartfiles -nostdlib -Wl,--defsym=STACK_SIZE=92160")
 
 # The target chip that is compiled for:
-include (${toolchain}/targets/arm32/cm3/atmel/sam3x/sam3x8e/sam3x8e.cmake)
+include (../../toolchain/targets/arm32/cm3/atmel/sam3x/sam3x8e/sam3x8e.cmake)
```
Comment (wvanbreukelen) [inline BuildModule.cmake]: Why would you hardcode this path?

File: src/main.cpp [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
+#include "mpu9250_interface.hpp"
+#include "navigation_engine.hpp"
+#include "vector2f.hpp"
 #include "wrap-hwlib.hpp"
 
 int main() {
     WDT->WDT_MR = WDT_MR_WDDIS;
+    auto scl = due::pin_oc(due::pins::scl);
+    auto sda = due::pin_oc(due::pins::sda);
+    const int8_t address = 0x68;
```
Comment (wvanbreukelen) [inline src/main.cpp]: Tip: could be a uint8_t

File: src/memory_map.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#ifndef MEMORY_MAP_HPP
+#define MEMORY_MAP_HPP
+
+/**
+ * @file        memory_map.hpp
+ * @brief       This header will contain memory addresses
```
Comment (wvanbreukelen) [inline src/memory_map.hpp]: I would note it as: This header will contain memory addresses for the MPU9250

File: src/mpu9250_interface.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * @author    Renske Kuip
+ * @license   See LICENSE
+ */
+
+#include "mpu9250_interface.hpp"
+#include "memory_map.hpp"
+
+void MPU9250Interface::init() {
+    hwlib::cout << hwlib::right << "Initializing MPU9250..." << hwlib::endl;
+    /*for (;;) {
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp]: Please remove debugging statements, or place them into a callable method.

File: src/mpu9250_interface.hpp [hunk 5]
```diff
@@ -8,17 +8,70 @@
...
-    MPU9250Interface() {
-    }
+    void init();
+
+    /**
+     * @brief Get the Accelerator values
+     *
+     * Returns values as array {x,y,z}
+     *
+     * @return int16_t Gyroscope values
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.hpp]: No return type

File: src/mpu9250_interface.hpp [hunk 6]
```diff
@@ -8,17 +8,70 @@
...
+     */
+    void getAccel();
+    /**
+     * @brief Get the Gyroscope values
+     *
+     * Returns values as array {x,y,z}
+     *
+     * @return int16_t Gyroscope values
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.hpp]: No return type

File: src/mpu9250_interface.hpp [hunk 7]
```diff
@@ -8,17 +8,70 @@
...
+     */
+    void getGyro();
+    /**
+     * @brief Get the Magnetometer values
+     *
+     * Returns values as array {x,y,z}
+     *
+     * @return int16_t Magnetometer values
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.hpp]: No return type

File: src/mpu9250_interface.hpp [hunk 8]
```diff
@@ -8,17 +8,70 @@
 #ifndef MPU9250_INTERFACE_HPP
 #define MPU9250_INTERFACE_HPP
 
+#include "wrap-hwlib.hpp"
+
 class MPU9250Interface {
   public:
+    MPU9250Interface(hwlib::pin_oc &scl, hwlib::pin_oc &sda, const int8_t MPUAddr)
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.hpp:15]: Tip: could be a ```uint8_t```

File: src/mpu9250_interface.hpp [hunk 9]
```diff
@@ -8,17 +8,70 @@
...
+     * @brief Prints values of the given int32_t array in the format "X:% Y:% Z:%"
+     *
+     * @param temp int32_t array with 3 values
+     */
+    void printValuesX_Y_Z(int32_t temp[3]);
+
+  private:
+    hwlib::pin_oc &scl;
+    hwlib::pin_oc &sda;
+    const int8_t MPUAddr;
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.hpp:66]: Tip: could be a ```uint8_t```

File: src/mpu9250_interface.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        getAccel();
+        getGyro();
+        getMagn();
+    }*/
+    uint8_t data[7] = {0x03}; ///< I2C slave 0 register address for AK8963  data
+    uint8_t magnReadFlag[1] = {0x0C | 0x80}, readData[1] = {0x87}, continious[1] = {0x06};
+
+    i2c.write(AK8963_CNTL1, continious, 1);           // start continious mode AK8963
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp]: Grammar issues within this section

File: src/mpu9250_interface.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    i2c.write(MPUREG_I2C_SLV0_ADDR, magnReadFlag, 1); // Set the I2C slave addres of AK8963 and set for read.
+    i2c.write(MPUREG_I2C_SLV0_REG, data, 1);          // I2C slave 0 register address from where to begin data transfe
+    i2c.write(MPUREG_I2C_SLV0_CTRL, readData, 1);     // Read 6 bytes from the magnetometer
+
+    hwlib::wait_us(10000);
+
+    uint8_t bypass[1] = {0x22};
+    i2c.write(INT_PIN_CFG, bypass, 1); // set bypass mode
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp]: Why don't directly place the byte into the write method. Like:
```cpp
i2c.write(INT_PIN_CFG, 0x22, 1); // set bypass mode
```

File: src/mpu9250_interface.cpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    uint8_t data[12] = {0x3B}; ///< address where MPU data starts
+
+    i2c.write(MPUAddr, data, 1);
+    i2c.read(MPUAddr, data, 12);
+
+    gyroValue[0] = (((int16_t)data[6] << 8) | (int16_t)data[7]);
+    gyroValue[1] = (((int16_t)data[8] << 8) | (int16_t)data[9]);
+    gyroValue[2] = (((int16_t)data[10] << 8) | (int16_t)data[11]);
+
+    printValuesX_Y_Z(gyroValue);
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp]: Please return the gyroValue array and print outside of your method.

File: src/mpu9250_interface.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        // will unlatch the data registers for the next measurement
+        uint8_t c = data[6]; // end data read by reading ST2
+
+        if (!(c & 0x08)) { // Check if magnetic sensor overflow set, if not then report data
+            magnValue[0] = ((int16_t)data[1] << 8) | (int16_t)data[0];
+            magnValue[1] = ((int16_t)data[3] << 8) | (int16_t)data[2];
+            magnValue[2] = ((int16_t)data[5] << 8) | (int16_t)data[4];
+        }
+    }
+    printValuesX_Y_Z(magnValue);
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp:69]: Please return the magnValue array and print outside of your method.

File: src/mpu9250_interface.cpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+void MPU9250Interface::printValuesX_Y_Z(int16_t temp[3]) {
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp:72]: You would need these methods if you print outside your class.

File: src/navigation_path.hpp [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+     * gets the Vector2f at the front of the current path.
+     *
+     * @return Vector2f
+     */
+    Vector2f getNextLocation();
+
+    /**
+     * @brief gets the length of the path.
+     *
+     * gets the current length of the path.
```
Comment (wvanbreukelen) [inline src/navigation_path.hpp]: I would just remove this extended documentation. It says exactly the same as the sentence above it.

File: src/navigation_path.hpp [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @return int
+     */
+    int getPathLength();
+
+  private:
+    Vector2f path[50];
```
Comment (wvanbreukelen) [inline src/navigation_path.hpp:53]: Tip: currently the path is limited to 50 items. Maybe you could make the class a template class and make it a template parameter?

File: src/main.cpp [hunk 17]
```diff
@@ -1,14 +1,20 @@
...
     NavigationEngine engine;
     IOInterface interface(engine);
+    MPU9250Interface sensor(scl, sda, address);
+    sensor.init();
     hwlib::wait_ms(1000);
-    Vector2f point1(1.5, 5.6);
-    hwlib::cout << int(point1.x) << ',' << int(point1.y) << hwlib::endl;
-    hwlib::cout << "HWLIB does not support floats.." << hwlib::endl;
+    engine.run();
+    return 0;
```
Comment (wvanbreukelen) [inline src/main.cpp:19]: Why would you return if the engine.run() method will run forever

File: src/mpu9250_interface.cpp [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    i2c.write(MPUREG_I2C_SLV0_CTRL, readData, 1);     // Read 6 bytes from the magnetometer
+
+    hwlib::wait_us(10000);
+
+    uint8_t bypass[1] = {0x22};
+    i2c.write(INT_PIN_CFG, bypass, 1); // set bypass mode
+}
+
+void MPU9250Interface::getAccel() {
+    uint8_t data[6] = {0x3B}; ///< address where MPU data starts
```
Comment (wvanbreukelen) [inline src/mpu9250_interface.cpp:29]: Why don't place this address within your memory map?

File: test/test_main.cpp [hunk 19]
```diff
@@ -1,6 +1,32 @@
 #define CATCH_CONFIG_MAIN // This tells Catch to provide a main() - only do this in one cpp file
+#include "../src/navigation_path.hpp"
 #include "catch.hpp"
 
 TEST_CASE("Example Test Case") {
```
Comment (wvanbreukelen) [inline test/test_main.cpp]: Please remove the example test case

File: src/navigation_path.hpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+ * @license   See LICENSE
+ */
+
+#ifndef NAVIGATION_PATH_HPP
+#define NAVIGATION_PATH_HPP
+
+#include "vector2f.hpp"
+#include <vector>
+
+class NavigationPath {
```
Comment (wvanbreukelen) [inline src/navigation_path.hpp]: Please start the ```@brief``` sections with a uppercase character, like: ```@brief Creates a NavigationPath object.```

File: src/navigation_path.cpp [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+}
+void NavigationPath::addPathLocation(Vector2f location) {
+    path[index] = location;
+    index++;
+    if (index >= 50) {
+        index = 0;
+    }
+}
+Vector2f NavigationPath::getNextLocation() {
+    auto nextLocation = path[head];
```
Comment (wvanbreukelen) [inline src/navigation_path.cpp:13]: I would prefer ```Vector2f``` instead of using ```auto```

File: src/navigation_path.hpp [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+ * @brief     NavigationPath Class
+ * @author    Quinten Perquin
+ * @license   See LICENSE
+ */
+
+#ifndef NAVIGATION_PATH_HPP
+#define NAVIGATION_PATH_HPP
+
+#include "vector2f.hpp"
+#include <vector>
```
Comment (wvanbreukelen) [inline src/navigation_path.hpp]: You cannot make use of <vector>, as it is disabled (heap not functioning).

File: src/navigation_engine.cpp [hunk 23]
```diff
@@ -0,0 +1,41 @@
+#include "navigation_engine.hpp"
+#include "navigation_path.hpp"
+#include "wrap-hwlib.hpp"
+#include <string>
```
Comment (wvanbreukelen) [inline src/navigation_engine.cpp]: You cannot make use of <string>, as it is disabled (heap not functioning).

**Review by wvanbreukelen — CHANGES_REQUESTED** [review]
Please see the placed comments. If you have any questions, feel free to ask.

**Review by wvanbreukelen — APPROVED** [review]

**Comment by wvanbreukelen** [comment]
Changes approved, but the Travis.CI test is failing, please contact the leads if your team cannot resolve it by themselve.

**Comment by elsavira** [comment]
Branch updated to include latest .fetch shell update. Tests now succeed.

**Comment by Michelbaartman** [comment]
Merging pull request as @wvanbreukelen has approved the changes.

**Pull request merged**